File: pymdurs/examples/building_from_ign.py
"""
Example: Load buildings from IGN API using rsmdu

This example demonstrates how to:
1. Create a BuildingCollection
2. Set a bounding box
3. Download and process buildings from IGN API
4. Convert to pandas DataFrame
"""

import sys

import pymdurs


def main():
    print("🏢 Loading buildings from IGN API...")

    # Create BuildingCollection (using alias created by rsmdu_helper)
    buildings = pymdurs.geometric.Building(
        output_path="./output", defaultStoreyHeight=3.0
    )

    # Set bounding box (La Rochelle area, France)
    # Format: min_x, min_y, max_x, max_y (WGS84, EPSG:4326)
    buildings.set_bbox(-1.152704, 46.181627, -1.139893, 46.18699)

    geo = buildings.geo_core
    print(f"📦 Bounding box set")
    print(f"📁 Output path: {geo.output_path}")

    # Run processing: downloads from IGN API and processes heights
    print("⏳ Downloading buildings from IGN API...")
    buildings = buildings.run()

    print(f"✅ Loaded {len(buildings)} buildings")

    # Convert to pandas DataFrame
    print("📊 Converting to pandas DataFrame...")
    df = buildings.to_pandas()

    print("\n📈 DataFrame info:")
    print(df.info())
    print("\n📊 First few rows:")
    print(df.head())
    print("\n📊 Statistics:")
    print(df.describe())

    return buildings, df


if __name__ == "__main__":
    buildings, df = main()

File: rsmdu/examples/building_from_ign.rs
/// Example: Using BuildingCollection with run() method
/// Following Python: buildings = Building(output_path='./')
///                   buildings.Bbox = [-1.152704, 46.181627, -1.139893, 46.18699]
///                   buildings = buildings.run()
use anyhow::Result;
use rsmdu::geometric::building::BuildingCollection;

fn main() -> Result<()> {
    println!("=== Example: BuildingCollection with run() ===\n");
    println!("Following Python code:");
    println!("  buildings = Building(output_path='./')");
    println!("  buildings.Bbox = [-1.152704, 46.181627, -1.139893, 46.18699]");
    println!("  buildings = buildings.run()\n");

    // Python: buildings = Building(output_path='./')
    // Following Python: def __init__(self, filepath_shp=None, output_path=None, defaultStoreyHeight=3.0, set_crs=None)
    let mut buildings = BuildingCollection::new(
        None,                         // filepath_shp (None = use IGN API)
        Some("./output".to_string()), // output_path
        3.0,                          // defaultStoreyHeight
        None,                         // set_crs (None = use default EPSG:2154)
    )?;

    // Python: buildings.Bbox = [-1.152704, 46.181627, -1.139893, 46.18699]
    // Set bounding box for IGN API request
    buildings.set_bbox(-1.152704, 46.181627, -1.139893, 46.18699)?;

    println!("Bounding box set:");
    println!("  - Longitude: -1.152704 to -1.139893");
    println!("  - Latitude: 46.181627 to 46.18699");
    println!("  - Area: La Rochelle, France");
    println!("  - Format: WGS84 (EPSG:4326)\n");

    // Python: buildings = buildings.run()
    // Following Python: def run(self)
    // - If filepath_shp is None: execute_ign(key="buildings") and load from GeoJSON
    // - Else: load from shapefile
    // - Process heights (fill missing heights, calculate mean, etc.)
    // - Return self
    println!("Running run()...");
    println!("  - Downloading from IGN API (key='buildings')");
    println!("  - Loading from GeoJSON");
    println!("  - Processing heights\n");

    let buildings = buildings.run()?;

    println!("Buildings loaded and processed: {}", buildings.len());

    if buildings.len() > 0 {
        println!(
            "\nMean height (area-weighted): {:.2} m",
            buildings.calculate_mean_height()
        );

        // Convert to Polars DataFrame (similar to to_gdf() in Python)
        let df = buildings.to_polars_df()?;
        println!("\nPolars DataFrame (equivalent to to_gdf() in Python):");
        println!("{}", df.head(Some(5)));
    }

    Ok(())
}
